---
layout:     post                    # 使用的布局（不需要改）
title:      .NET 线程池的使用问题(2)             # 标题 
subtitle:   线程池模式下的编程实例 #副标题
date:       2020-03-13              # 时间
author:     Peter                      # 作者
header-img: img/post-bg-2015.jpg    #这篇文章标题背景图片
catalog: true                       # 是否归档
tags:                               #标签
    - .NET
    - Parallel
---

本篇会介绍在线程池模式下取消以及设置等待事件等相关操作的实例。

## 取消线程

在线程池的编程中会使用[CancellationToken](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=netframework-4.8)来协同多线程任务中特定线程的取消操作。  
具体的，通过初始化[CancellationTokenSource](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource?view=netframework-4.8)来获得一个新的`CancellationToken`。  
```
    using (var cts = new CancellationTokenSource())
    {
        CancellationToken token = cts.Token;
        // Use token
         cts.Cancel();
    }
```
具体实践中，通常通过三种方法实现线程的取消操作：

### 1. 通过轮询检查`IsCancellationRequested`参数

代码运行时轮询token的`IsCancellationRequested`来判断是否需要取消当前任务。 

```
    static void AsyncOperation1(CancellationToken token)
    {
        WriteLine("Starting the first task");
        for (int i = 0; i < 5; i++)
        {
            if (token.IsCancellationRequested)
            {
                WriteLine("The first task has been canceled.");
                return;
            }
            Sleep(TimeSpan.FromSeconds(1));
        }
        WriteLine("The first task has completed succesfully");
    }
```

### 2. 通过抛出`ThrowIfCancellationRequested`异常

代码通过抛出`OperationCanceledException`来取消操作。这往往需要在具体的线程中判断是否需要取消当前的线程，并准备对应的try/catch机制。  

```
    static void AsyncOperation2(CancellationToken token)
    {
        try
        {
            WriteLine("Starting the second task");

            for (int i = 0; i < 5; i++)
            {
                token.ThrowIfCancellationRequested();
                Sleep(TimeSpan.FromSeconds(1));
            }
            WriteLine("The second task has completed succesfully");
        }
        catch (OperationCanceledException)
        {
            WriteLine("The second task has been canceled.");
        }
    }
```

### 3. 通过注册一个回调函数

通过`CancellationToken.Register()`注册一个取消时的回调函数并配合轮询的判断逻辑来通过线程取消。

```
    static void AsyncOperation3(CancellationToken token)
    {
        bool cancellationFlag = false;
        token.Register(() => cancellationFlag = true);
        WriteLine("Starting the third task");
        for (int i = 0; i < 5; i++)
        {
            if (cancellationFlag)
            {
                WriteLine("The third task has been canceled.");
                return;
            }
            Sleep(TimeSpan.FromSeconds(1));
        }
        WriteLine("The third task has completed succesfully");
    }
```


## 使用等待事件处理器以及超时

线程池会使用`ThreadPool.RegisterWaitForSingleObject()`方法来管理timeout问题。  
`ThreadPool.RegisterWaitForSingleObject()`主要会接受
1. 一个WaitHanle；
2. 一个处理超时的回调函数Callback
3. 一个timeout时间

CallBack使用是的delegate `WaitOrTimerCallback`

```
[System.Runtime.InteropServices.ComVisible(true)]
public delegate void WaitOrTimerCallback(object state, bool timedOut);
```

有两种情况会触发处理超时的异步操作：
1. WaitHandle被触发，此时timeout还没有计时结束, timeOut = false;
2. Timeout结束，此时WaitHandle还没有被触发, timeOut = true.

```
    static void RunOperations(TimeSpan workerOperationTimeout)
    {
        using (var evt = new ManualResetEvent(false))
        using (var cts = new CancellationTokenSource())
        {
            WriteLine("Registering timeout operation...");
            var worker = ThreadPool.RegisterWaitForSingleObject(evt
                , (state, isTimedOut) => WorkerOperationWait(cts, isTimedOut)
                , null
                , workerOperationTimeout
                , true);

            worker.Unregister(evt);
        }

    }
```
上述示例采用`ManualResetEvent`作为一个WaitHandle, 可以通过`ManualResetEvent.Set()`来手动触发事件。  
WaitHandle一般放置在具体需要处理的业务逻辑中，并在处理结束后`Set()`发出信号回调CallBack()。  




